init/nitro/main.c

```diff
@@ -177,32 +181,118 @@ krun_signal()
     // communication is established.
     ret = write(sock_fd, buf, 1);
     if (ret != 1) {
+        close(sock_fd);
         perror("vsock write");
         return -errno;
     }
 
     ret = read(sock_fd, buf, 1);
     if (ret != 1) {
+        close(sock_fd);
         perror("vsock read");
         return -errno;
     }
 
-    if (buf[0] != HEART_BEAT)
+    if (buf[0] != HEART_BEAT) {
+        close(sock_fd);
+        return -1;
+    }
+
+    return sock_fd;
+}
+
+static int
+rootfs_rcv(int sock_fd, void **buf_ptr, size_t *size)
+{
+    uint8_t len_b[8], *rootfs_archive;
+    ssize_t read_len;
+    uint64_t len, idx = 0;
+
+    read_len = read(sock_fd, len_b, 8);
+    if (read_len < 8) {
+        close(sock_fd);
+        perror("vsock rootfs archive length read");
         return -1;
+    }
+
+    memcpy(&len, len_b, sizeof(uint64_t));
+
+    *size = len;
```

Comment:
need to check if size is NULL first?

Reply:
I don't think so. size is allocated on the stack from the caller. If this was a library meant to be used by other consumers, sure.